# Progetto data science e pensiero computazionale <br>
# A.A. 2025/2026<br>
# Fase 1 - descrizione del dataset

Requisiti - almeno:

*   8 domande/ipotesi
*   5 statistiche descrittive;
*   3 riflessioni critiche sui dati.

Esempi:


*   class imbalance;
*   outlier;
*   feature correlate;
*   dati mancanti;
*   feature sospette.

## Preparazione dell'ambiente

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/elisamezzi/Progetto_IDSAPCProg_gruppo5/main/data/Employee_Attrition.csv"
df_EA = pd.read_csv(url)

#Chiesto a LLM come bypassare il problema del recupero del dataset colab - github, letto via URL raw

##Analisi descrittiva del dataset

In [2]:
num_righe = df_EA.shape[0]
num_colonne = df_EA.shape[1]
print("EMPLOYEE ATTRITITION DATASET")
print("Numero righe", num_righe)
print("Numero colonne", num_colonne)
print(" ")

EMPLOYEE ATTRITITION DATASET
Numero righe 1470
Numero colonne 35
 


In [3]:
#verifico presenza valori nulli nel dataset.
print(df_EA.isnull().sum()[df_EA.isnull().sum() > 0])

#-> dataset pulito

Series([], dtype: int64)


Come sono divise le variabili?

In [4]:
df_EA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

Divido la tipologia di colonne (numeriche, categoriche pure, categoriche codificate)

In [7]:
numeriche = df_EA.select_dtypes(include=['int64', 'float64']).columns.tolist() #trova tutte le colonne che hanno dtype int64 o float64 nel dataframe, e le salva in una lista chiamata numeriche.

#divido in sottoinsiemi le categoriche in codificate e pure (in base a file description)
cat_pure = ['Attrition', 'Gender', 'Department', 'JobRole',
                'MaritalStatus', 'BusinessTravel', 'EducationField',
                'OverTime', 'Over18']

cat_codificate = ['Education', 'EnvironmentSatisfaction', 'JobInvolvement',
                'JobSatisfaction', 'PerformanceRating',
                'RelationshipSatisfaction', 'WorkLifeBalance']

numeriche_pure = [col for col in numeriche if col not in cat_codificate]
#verifico visivamente che le tre liste siano corrette
display(df_EA[numeriche_pure].head())
print()
display(df_EA[cat_pure].head())
print()
display(df_EA[cat_codificate].head())


,Age,DailyRate,DistanceFromHome,EmployeeCount,EmployeeNumber,HourlyRate,JobLevel,MonthlyIncome,MonthlyRate,NumCompaniesWorked,PercentSalaryHike,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,1102,1,1,1,94,2,5993,19479,8,11,80,0,8,0,6,4,0,5
1,49,279,8,1,2,61,2,5130,24907,1,23,80,1,10,3,10,7,1,7
2,37,1373,2,1,4,92,1,2090,2396,6,15,80,0,7,3,0,0,0,0
3,33,1392,3,1,5,56,1,2909,23159,1,11,80,0,8,3,8,7,3,0
4,27,591,2,1,7,40,1,3468,16632,9,12,80,1,6,3,2,2,2,2


,Attrition,Gender,Department,JobRole,MaritalStatus,BusinessTravel,EducationField,OverTime,Over18
0,Yes,Female,Sales,Sales Executive,Single,Travel_Rarely,Life Sciences,Yes,Y
1,No,Male,Research & Development,Research Scientist,Married,Travel_Frequently,Life Sciences,No,Y
2,Yes,Male,Research & Development,Laboratory Technician,Single,Travel_Rarely,Other,Yes,Y
3,No,Female,Research & Development,Research Scientist,Married,Travel_Frequently,Life Sciences,Yes,Y
4,No,Male,Research & Development,Laboratory Technician,Married,Travel_Rarely,Medical,No,Y


,Education,EnvironmentSatisfaction,JobInvolvement,JobSatisfaction,PerformanceRating,RelationshipSatisfaction,WorkLifeBalance
0,2,2,3,4,3,1,1
1,1,3,2,2,4,4,3
2,2,4,2,3,3,2,3
3,4,4,3,3,3,3,3
4,1,1,3,2,3,4,3


Come sono distribuite le variabili numeriche?

In [12]:
#valori minimi massimi e medie var numeriche
tab_generale_numeriche = pd.DataFrame({
    "media": df_EA[numeriche_pure].mean(),
    "mediana": df_EA[numeriche_pure].median(),
    "min": df_EA[numeriche_pure].min(),
    "max": df_EA[numeriche_pure].max()
})
print(tab_generale_numeriche)
tab_generale_numeriche.to_html("tab_generale_numeriche.html", border=1)

                                media  mediana   min    max
Age                         36.923810     36.0    18     60
DailyRate                  802.485714    802.0   102   1499
DistanceFromHome             9.192517      7.0     1     29
EmployeeCount                1.000000      1.0     1      1
EmployeeNumber            1024.865306   1020.5     1   2068
HourlyRate                  65.891156     66.0    30    100
JobLevel                     2.063946      2.0     1      5
MonthlyIncome             6502.931293   4919.0  1009  19999
MonthlyRate              14313.103401  14235.5  2094  26999
NumCompaniesWorked           2.693197      2.0     0      9
PercentSalaryHike           15.209524     14.0    11     25
StandardHours               80.000000     80.0    80     80
StockOptionLevel             0.793878      1.0     0      3
TotalWorkingYears           11.279592     10.0     0     40
TrainingTimesLastYear        2.799320      3.0     0      6
YearsAtCompany               7.008163   

In [9]:
# Distribuzione delle varibili numeriche in quantili + min e max e media (utile x outlier)

distribuzione = pd.DataFrame({
    "min": df_EA[numeriche_pure].min(),
    "25%": df_EA[numeriche_pure].quantile(0.25),
    "50%": df_EA[numeriche_pure].quantile(0.50),
    "75%": df_EA[numeriche_pure].quantile(0.75),
    "max": df_EA[numeriche_pure].max(),
    "mean": df_EA[numeriche_pure].mean()
})

print(distribuzione)

                          min      25%      50%       75%    max          mean
Age                        18    30.00     36.0     43.00     60     36.923810
DailyRate                 102   465.00    802.0   1157.00   1499    802.485714
DistanceFromHome            1     2.00      7.0     14.00     29      9.192517
EmployeeCount               1     1.00      1.0      1.00      1      1.000000
EmployeeNumber              1   491.25   1020.5   1555.75   2068   1024.865306
HourlyRate                 30    48.00     66.0     83.75    100     65.891156
JobLevel                    1     1.00      2.0      3.00      5      2.063946
MonthlyIncome            1009  2911.00   4919.0   8379.00  19999   6502.931293
MonthlyRate              2094  8047.00  14235.5  20461.50  26999  14313.103401
NumCompaniesWorked          0     1.00      2.0      4.00      9      2.693197
PercentSalaryHike          11    12.00     14.0     18.00     25     15.209524
StandardHours              80    80.00     80.0     

Come sono distribuite le variabili categoriche "pure"?

In [10]:
distribuzioni = {}
for col in cat_pure:
    distribuzioni[col] = df_EA[col].value_counts()

print("= VARIABILI CATEGORICHE PURE =\n")
for col, valori in distribuzioni.items():
    totale = valori.sum()
    tabella = pd.DataFrame({
        "count": valori,
        "%": (valori / totale * 100).round(1)
    })
    print(f"- {col} -")
    print(tabella.to_string())
    print()

= VARIABILI CATEGORICHE PURE =

- Attrition -
           count     %
Attrition             
No          1233  83.9
Yes          237  16.1

- Gender -
        count     %
Gender             
Male      882  60.0
Female    588  40.0

- Department -
                        count     %
Department                         
Research & Development    961  65.4
Sales                     446  30.3
Human Resources            63   4.3

- JobRole -
                           count     %
JobRole                               
Sales Executive              326  22.2
Research Scientist           292  19.9
Laboratory Technician        259  17.6
Manufacturing Director       145   9.9
Healthcare Representative    131   8.9
Manager                      102   6.9
Sales Representative          83   5.6
Research Director             80   5.4
Human Resources               52   3.5

- MaritalStatus -
               count     %
MaritalStatus             
Married          673  45.8
Single           470  32.0
Divor

In [14]:
#frequenza delle categoriche pure
righe = []
for col in cat_pure:
    valori = df_EA[col].value_counts()
    righe.append({
        "variabile": col,
        "categoria_piu_frequente": valori.index[0],
        "count_piu_frequente": valori.iloc[0],
        "%_piu_frequente": round(valori.iloc[0] / len(df_EA) * 100, 1)
    })

tab_generale_categoriche = pd.DataFrame(righe)
print(tab_generale_categoriche)

        variabile categoria_piu_frequente  count_piu_frequente  \
0       Attrition                      No                 1233   
1          Gender                    Male                  882   
2      Department  Research & Development                  961   
3         JobRole         Sales Executive                  326   
4   MaritalStatus                 Married                  673   
5  BusinessTravel           Travel_Rarely                 1043   
6  EducationField           Life Sciences                  606   
7        OverTime                      No                 1054   
8          Over18                       Y                 1470   

   %_piu_frequente  
0             83.9  
1             60.0  
2             65.4  
3             22.2  
4             45.8  
5             71.0  
6             41.2  
7             71.7  
8            100.0  


Come sono distribuite le variabili categoriche codificate?

In [18]:
righe = []
for col in cat_codificate:
    valori = df_EA[col].value_counts().sort_index()
    righe.append({
        "variabile": col,
        "val_piu_frequente": valori.idxmax(), #idxmax() invece di index[0] perché le codificate sono ordinate per val numerico e non per frequenza -> idxmax() trova direttamente il valore più frequente.
        "%_piu_frequente": round(valori.max() / len(df_EA) * 100, 1)
    })

tab_generale_codificate = pd.DataFrame(righe)
print(tab_generale_codificate)

                  variabile  val_piu_frequente  %_piu_frequente
0                 Education                  3             38.9
1   EnvironmentSatisfaction                  3             30.8
2            JobInvolvement                  3             59.0
3           JobSatisfaction                  4             31.2
4         PerformanceRating                  3             84.6
5  RelationshipSatisfaction                  3             31.2
6           WorkLifeBalance                  3             60.7


##Statistiche descrittive
1 - Il dataset contiene 1470 dipendenti con età compresa tra 18 e 60 anni (media 37).<br>
2 - Il reddito mensile varia molto: da 1009\$ a 19999\$, con una media di 6503\$ e forte asimmetria (metà dei dipendenti guadagna meno di 4919\$).<br>
3 - `Attrition`(variabile target!):
presupponendo che attrition (logoramento) = abbandono/ dimissioni, la percentuale di chi ha lasciato il posto di lavoro è molto bassa (16.1%) e questo sbilancia il dataset.<br>
4 - `OverTime`: Quasi il 30% dei dipendenti fa straordinari. <br>
5 - `Department`: Il 65% dei dipendenti è in R&D, solo il 4.3% in HR. <br>
6 - `BusinessTravel`: Il 71% viaggia raramente, il 18.8% frequentemente. Il 10% per nulla.<br>
7 - `Education`: la maggior parte dei dipendenti ha un livello Bachelor (38.9%) o Master (27.1%). I Doctor sono pochissimi (3.3%).<br>
8 - `JobInvolvement`il 59% dei dipendenti ha coinvolgimento "High" (3). La distribuzione è asimmetrica verso l'alto.<br>
9 - `WorkLifeBalance` similmente, il 60.7% è al livello 3 ("Better"). Pochissimi (5.4%) riportano un work-life balance "Bad".<br>

10 - `PerformanceRating`: vera anomalia di distribuzione. Sono valorizzati solo 3 e 4, 1 e 2 non sono presenti nel dataset. Vuol dire che tutti i dipendenti hanno performance "Excellent" o "Outstanding". Perchè? (LLM suggerisce che i dipendenti con rating basso siano già stati rimossi dal dataset prima della raccolta)<br>


**Importante per la fase di modellazione:**<br>
- `Over18`: il 100% dei dipendenti è maggiorenne;
- `EmployeeCount` vale sempre 1;
- `StandardHours`: vale sempre 80;
- `EmployeeNumber`: sembrerebbe solo ID del dipendente.
<br>
Valutare se eliminarle dal DF.

##Domande o ipotesi sul dataset

1 - I dipendenti con basso `JobSatisfaction` abbandonano più spesso?<br>
2 - Lo stipendio `MonthlyIncome` incide sull'abbandono?<br>
3 - I giovani abbandonano più dei senior?<br>
4 - I dipendenti che fanno straordinari abbandonano più spesso?<br>
5 - Il numero di aziende in cui si è lavorato in passato `NumCompaniesWorked` è correlato all'abbandono?<br>
6 - Qual è il rapporto tra promozioni `YearsSinceLastPromotion` e `Attritition`? Il tempo che passa tra una promozione e la successiva incide sull'abbandono?<br>
7 - la frequenza dei viaggi di lavoro incide sull'abbandono? <br>
8 - Il reparto in cui si lavora influenza la probabilità di lasciare?<br>



In [24]:
#Attrition vs variabili categoriche

for col in ['JobSatisfaction', 'OverTime', 'BusinessTravel','Department']:
    tabella = df_EA.groupby(col)['Attrition'].value_counts(normalize=True).mul(100).round(1).unstack()
    print(f"- {col} -")
    print(tabella)
    print()



- JobSatisfaction -
Attrition          No   Yes
JobSatisfaction            
1                77.2  22.8
2                83.6  16.4
3                83.5  16.5
4                88.7  11.3

- OverTime -
Attrition    No   Yes
OverTime             
No         89.6  10.4
Yes        69.5  30.5

- BusinessTravel -
Attrition            No   Yes
BusinessTravel               
Non-Travel         92.0   8.0
Travel_Frequently  75.1  24.9
Travel_Rarely      85.0  15.0

- Department -
Attrition                 No   Yes
Department                        
Human Resources         81.0  19.0
Research & Development  86.2  13.8
Sales                   79.4  20.6



In [22]:
# Attrition vs variabili numeriche

for col in ['MonthlyIncome', 'Age', 'NumCompaniesWorked', 'YearsSinceLastPromotion']:
    tabella = df_EA.groupby('Attrition')[col].mean().round(1)
    print(f"- {col} -")
    print(tabella)
    print()

- MonthlyIncome -
Attrition
No     6832.7
Yes    4787.1
Name: MonthlyIncome, dtype: float64

- Age -
Attrition
No     37.6
Yes    33.6
Name: Age, dtype: float64

- NumCompaniesWorked -
Attrition
No     2.6
Yes    2.9
Name: NumCompaniesWorked, dtype: float64

- YearsSinceLastPromotion -
Attrition
No     2.2
Yes    1.9
Name: YearsSinceLastPromotion, dtype: float64



1 - `JobSatisfaction`: sì, chi è molto insoddisfatto abbandona il doppio rispetto a chi è molto soddisfatto. <br>
2 - `MonthlyIncome`: sì, chi guadagna meno (in media 4787\$) tende ad andarsene rispetto a chi guadagna di più (in media 6833\$).<br>
3 - `Age`: sì, i dipendenti più giovani tendono ad abbandonare di più(33.6 anni vs 37.6).<br>
4 - `OverTime`: chi fa straordinari abbandona il 30.5% delle volte, chi non li fa solo il 10.4%.<br>
5 - `NumCompaniesWorked`: poca differenza tra chi ha cambiato più aziende e chi meno (2.6 vs 2.9).<br>
6 - `YearsSinceLastPromotion`: chi ha abbandonato ha aspettato meno per una promozione rispetto a chi è rimasto (la nostra ipotesi era quella opposta!).<br>
7 - `BusinessTravel`: chi viaggia frequentemente abbandona quasi il 25% delle volte, chi non viaggia mai solo l'8%. <br>
8 - `Department`: sì, il reparto sembra influenzare l'abbandono (Sales e HR sono rispettivamente al 20.6% e 19%, R&D al 13.8%).

##Riflessioni sui dati

Gli straordinari sono il segnale più forte — con il 30.5% di abbandono tra chi fa overtime contro il 10.4% di chi non ne fa, è la variabile con la differenza più notevole tutto il dataset. <br><br>

Anche chi è abbastanza soddisfatto del lavoro (livello 2 e 3) abbandona comunque intorno al 16%, quasi quanto chi è insoddisfatto. Quindi la soddisfazione non sembra poter trattenere i dipendenti da sola. Lo stipendio invece mostra una differenza molto più marcata e netta tra i due gruppi. Quindi i fattori economici sembrerebbe che pesino di più di quelli di soddisfazione nella decisione di cambiare lavoro.<br><br>

Su `YearsSinceLastPromotion` la nostra ipotesi era quella opposta a quello che i dati mostrano; ci aspettavamo che chi aspetta più a lungo una promozione fosse più propenso ad andarsene, ma è il contrario. <br><br>

`PerformanceRating` è una variabile incompleta: nel dataset esistono solo i valori 3 (Excellent) e 4 (Outstanding), mentre 1 (Low) e 2 (Good) sono completamente assenti. Di conseguenza questa variabile non può essere usata in modo affidabile per predire l'abbandono perchè chi abbandona per scarsa performance è già stato escluso dal campione.